waiter/src/waiter/reporter.clj

```diff
@@ -0,0 +1,151 @@
+
+  (state [this]
+    "Returns the state of this reporter"))
+
+(defn- make-metrics-filter
+  [filter-regex]
+  (reify MetricFilter (matches [_ name _] (some? (re-matches filter-regex name)))))
+
+(defn- make-codahale-reporter
+  [^ScheduledReporter scheduled-reporter state period-ms]
+  (.start scheduled-reporter period-ms TimeUnit/MILLISECONDS)
+  (reify CodahaleReporter
+    (close! [_] (.close scheduled-reporter) (reset! state :closed))
+    (state [_] @state)))
+
+
+(defn validate-console-reporter-config
+  "Validates ConsoleReporter settings and sets defaults"
+  [config]
+  (s/validate {(s/required-key :period-ms) schema/positive-int
+               (s/optional-key :filter-regex) s/Regex
+               s/Any s/Any} config))
+
+(defn make-console-reporter
+  "Creates a ConsoleReporter for codahale metrics"
+  ([filter-regex] (make-console-reporter filter-regex nil))
+  ([filter-regex ^PrintStream output]
+   (let [state (atom {})
+         console-reporter (cond-> (ConsoleReporter/forRegistry metrics/default-registry)
+                                  output (.outputTo output)
+                                  filter-regex (.filter (make-metrics-filter filter-regex))
+                                  true (-> (.convertRatesTo TimeUnit/SECONDS)
+                                           (.convertDurationsTo TimeUnit/MILLISECONDS)
+                                           (.build)))]
+     [console-reporter state])))
+
+(defn console-reporter
+  "Creates and starts a ConsoleReporter for codahale metrics"
+  [config]
+  (let [{:keys [period-ms filter-regex]} (validate-console-reporter-config config)
+        [console-reporter state] (make-console-reporter filter-regex)]
+    (make-codahale-reporter console-reporter state period-ms)))
+
+
+
+(defn validate-graphite-reporter-config
+  "Validates GraphiteReporter settings and sets defaults"
+  [config]
+  (s/validate {(s/required-key :period-ms) schema/positive-int
+               (s/optional-key :filter-regex) s/Regex
+               (s/required-key :host) s/Str
+               (s/optional-key :prefix) s/Str
+               (s/required-key :port) schema/positive-int
+               (s/required-key :pickled?) s/Bool
+               s/Any s/Any}
+              (merge {:pickled? true} config)))
+
+(defn make-graphite-reporter
+  "Creates a GraphiteReporter for metrics"
+  ([filter-regex prefix host port pickled?]
+   (let [addr (InetSocketAddress. ^String host ^int port)
+         ^GraphiteSender graphite (if pickled?
+                                    (PickledGraphite. addr)
+                                    (Graphite. addr))]
+     (make-graphite-reporter filter-regex prefix graphite)))
+  ([filter-regex prefix ^GraphiteSender graphite]
+   (let [state (atom {})
+         graphite-wrapper (reify GraphiteSender
+                            (connect [_]
+                              (try (.connect graphite)
+                                   (catch Exception e
```

Comment:
Should this exception be logged?

Reply:
Logged in the rest of the stack. Would it be double logged? Should I not rethrow?

Reply:
The `connect` javadoc says it expects the exceptions to be thrown if there is an error.
Where is the exception logged? We should logging something in our code, it does not need to include the stack trace if it is available from the other log. Can you include an example log?